### DISCUSSION TITLE
Integration with QuickGrid

### DISCUSSION THREAD

**Discussion by mdmontesinos** [oqtane/oqtane.framework#4102] [Q&A]
Has anyone being able to integrate the new asp.net core official [QuickGrid ](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/quickgrid?view=aspnetcore-8.0) component in Oqtane?

I've installed the Nuget package in both client and server, copied the dll in the Package project, extracted the staticwebassets into _content/Microsoft.AspNetCore.Components.QuickGrid and referenced them from my module resources. 

However, the issue is that the js for QuickGrid is scoped to the razor component "QuickGrid.razor.js" which throws a compile error in both Oqtane.Server and my Server project:

"Error BLAZOR106 The JS module file '~\oqtane.framework\Oqtane.Server\wwwroot\ _content\Microsoft.AspNetCore.Components.QuickGrid\QuickGrid.razor.js' was defined but no associated razor component or view was found for it."

If you try to rename it to QuickGrid.js, it compiles but you end up getting an error from the QuickGrid razor component not being able to find the javascript file.

```
Microsoft.JSInterop.JSException: Failed to fetch dynamically imported module: http://localhost:44357/_content/Microsoft.AspNetCore.Components.QuickGrid/QuickGrid.razor.js
TypeError: Failed to fetch dynamically imported module: http://localhost:44357/_content/Microsoft.AspNetCore.Components.QuickGrid/QuickGrid.razor.js
   at Microsoft.JSInterop.JSRuntime.InvokeAsync[TValue](Int64 targetInstanceId, String identifier, Object[] args)
   at Microsoft.AspNetCore.Components.QuickGrid.QuickGrid`1.OnAfterRenderAsync(Boolean firstRender)
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.GetErrorHandledTask(Task taskToHandle, ComponentState owningComponentState)
```

**Comment by sbwalker** [comment]
@mdmontesinos can you share your Resources declaration in your module which references the QuickGrid JavaScript file.

**Reply by mdmontesinos** [reply]
There it is:

```c#
Resources =
[
    new Resource { ResourceType = ResourceType.Stylesheet, Url = "~/Module.css" },
    new Resource { ResourceType = ResourceType.Script, Url = "~/Module.js", Location = ResourceLocation.Body },
    new Resource { ResourceType = ResourceType.Stylesheet, Url = "_content/Microsoft.AspNetCore.Components.QuickGrid/Microsoft.AspNetCore.Components.QuickGrid.bundle.scp.css" },
    new Resource { ResourceType = ResourceType.Script, ES6Module = true, Url = "_content/Microsoft.AspNetCore.Components.QuickGrid/QuickGrid.razor.js", Location = ResourceLocation.Body }
]
```

I've obviously also tried the version for QuickGrid.js, removing the .razor part.

**Reply by sbwalker** [reply]
@mdmontesinos I have not used QuickGrid... originally it was an experimental component created by Steve Sanderson to demonstrate how to do high performance pagination (it follows a similar approach as the Oqtane Pager ie. avoiding micro-components, etc...). In .NET 8 I believe it is shipped as a first class citizen in the .NET framework - or am I mistaken?

**Reply by mdmontesinos** [reply]
Yes, it is shipped in .NET8 as part of the AspNetCore github repo but you still need to install it as a NuGet package:

![image](https://github.com/oqtane/oqtane.framework/assets/90258222/ee52bb67-b2b1-46b0-a814-3d25a490674c)

**Reply by sbwalker** [reply]
Ah I see that now. Here is the source code for the QuickGrid component:

https://github.com/dotnet/aspnetcore/blob/2a04de4e70e36a37b9042f2ae0d6e9b9aa0c0ed2/src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/QuickGrid.razor.cs

I see the following hardcoded references to the JavaScript file:

```
    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        if (firstRender)
        {
            _jsModule = await JS.InvokeAsync<IJSObjectReference>("import", "./_content/Microsoft.AspNetCore.Components.QuickGrid/QuickGrid.razor.js");
            _jsEventDisposable = await _jsModule.InvokeAsync<IJSObjectReference>("init", _tableReference);
        }
```

**Reply by sbwalker** [reply]
Based on the above, you should not include a reference to the JS file in your Resources declaration - as the component is taking care of regitering itself. You would need to include the JS file in the hard-coded location specified and you would need to include the assembly in the /bin.

**Reply by mdmontesinos** [reply]
The thing is that I can't include it in the _content folder because it's a scoped js file and it can't find the associated razor component. Visual Studio throws a compilation error:

"Error BLAZOR106 The JS module file 'MY_PATH\oqtane.framework\Oqtane.Server\wwwroot\ _content\Microsoft.AspNetCore.Components.QuickGrid\QuickGrid.razor.js' was defined but no associated razor component or view was found for it."

**Reply by sbwalker** [reply]
I do not understand the error - if the QuickGrid DLL is in your /bin folder then why can't it find the razor component?

**Reply by sbwalker** [reply]
Other people having difficulty with this: https://github.com/dotnet/aspnetcore/issues/39344#issuecomment-1982209824

**Reply by mdmontesinos** [reply]
Just for clarification, here's how I copied the dll:

```
XCOPY "..\MyProject.Server\bin\Debug\net8.0\Microsoft.AspNetCore.Components.QuickGrid.dll" "..\..\oqtane.framework\Oqtane.Server\bin\Debug\net8.0\" /Y
```

And it is indeed copied:
![image](https://github.com/oqtane/oqtane.framework/assets/90258222/ddd8f44c-5e86-4f7c-8d54-75d93a9197ca)

**Reply by sbwalker** [reply]
I am guessing it is because Visual Studio (and Blazor to a large extent) relies on tight coupling where all app dependencies are declared directly in the project file. This allows it to do "magic" at deployment as it has all of the context it needs to deploy files to the appropriate locations. However, tight coupling results in monolithic apps and monolithic deployments.... which is why Oqtane is modular and uses loose coupling. But Visual Studio always tries to fight back.

**Reply by mdmontesinos** [reply]
Is there any way that you know to suppress the error and allow Visual Studio to build the project?

Otherwise, I'll log an issue at the asp repo to see if they come up with a solution.

**Reply by sbwalker** [reply]
Is it a Visual Studio error or a .NET compiler error?

**Reply by sbwalker** [reply]
See this: https://stackoverflow.com/questions/73126368/load-js-file-with-custom-razor-component-in-blazor-app

"Fixed by marking component as Content instead of None"

To answer your question above I have never seen this error before so I do not have a solution. I found the above suggestion by Googling "Error BLAZOR106" - and there are a LOT of references to it online

**Reply by mdmontesinos** [reply]
`Is it a Visual Studio error or a .NET compiler error?`
I'd say a .NET compiler error, as it throws when trying to build the solution in both the Oqtane and Module Server projects,
```
1>------ Build started: Project: Oqtane.Server, Configuration: Debug Any CPU ------
2>------ Build started: Project: MyModule.Server, Configuration: Debug Any CPU ------
2>PATH\MyModule.Server\wwwroot\_content\Microsoft.AspNetCore.Components.QuickGrid\QuickGrid.razor.js : error BLAZOR106: The JS module file 'PATH\MyModule.Server\wwwroot\_content\Microsoft.AspNetCore.Components.QuickGrid\QuickGrid.razor.js' was defined but no associated razor component or view was found for it.
2>Done building project "MyModule.Server.csproj" -- FAILED.
1>PATH\oqtane.framework\Oqtane.Server\wwwroot\_content\Microsoft.AspNetCore.Components.QuickGrid\QuickGrid.razor.js : error BLAZOR106: The JS module file 'PATH\oqtane.framework\Oqtane.Server\wwwroot\_content\Microsoft.AspNetCore.Components.QuickGrid\QuickGrid.razor.js' was defined but no associated razor component or view was found for it.
1>Done building project "Oqtane.Server.csproj" -- FAILED.
```

The error list also permanently displays this error also associated with the QuickGrid js file:
```
Severity	Code	Description	Project	File	Line	Suppression State	Details
Error	TS6053	File 'PATH/MyModule/wwwroot/_content/Microsoft.AspNetCore.Components.QuickGrid/QuickGrid.razor.js' not found.
  The file is in the program because:
    Root file specified for compilation	Oqtane.Server JavaScript Content Files	PATH\MyModule\MyModule.sln	1	Active	
```

And about `"Fixed by marking component as Content instead of None"`, I don't actually have the QuickGrid component, it's a referenced DLL, so I'm not entirely sure how that's achievable.

I've also search for the BLAZOR106 error, but found nothing useful, I might log an issue in the asp repo. Btw, how do you usually provide minimal repos or explain errors to the ASP/.NET team when they are related to using Oqtane?

**Reply by sbwalker** [reply]
@mdmontesinos I usually create a default Blazor application using the template and try to reproduce the problem - but it is often very difficult to create a simple reproduction (I have spent days just trying to create a simple repro before).

**Comment by mdmontesinos** [comment]
Reported the issue in aspnetcore repo: https://github.com/dotnet/aspnetcore/issues/55009

**Comment by mdmontesinos** [comment]
The asp.net core team just answered to the issue I posted about integrating QuickGrid.

Basically, there's no "good" way to integrate this component in Oqtane, as it would either require modifying MSBuild to supress the BLAZOR106 error or creating an importmap so that I can rename the scoped file to QuickGrid.js (without the .razor) and Quickgrid is still able to load it.

Also, @sbwalker they seem to be "critical" with the way Oqtane integrates external RCLs.

> The issue seems to stem from the way Oqtane operates by bypassing the usage of static web assets all together. It might be ok for Oqtane not to integrate in that way, but then its on them the responsibility to adjust the build as necessary to make this work. I would go even further and say that this is unsupported territory from a product point of view and might break at any moment in the future.

It's also possible I didn't explained properly how Oqtane operates, as I'm still getting the grasp for it. Perhaps you'd like to weigh in the conversation.
